src/GMCPAuthenticator.cpp

```diff
@@ -0,0 +1,127 @@
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
+ ***************************************************************************/
+
+#include "GMCPAuthenticator.h"
+
+#include "Host.h"
+#include "ctelnet.h"
+#include <QDebug>
+
+GMCPAuthenticator::GMCPAuthenticator(Host* pHost)
+: mpHost(pHost)
+{}
+
+void GMCPAuthenticator::saveSupportsSet(const QString& data)
+{
+    auto jsonDoc = QJsonDocument::fromJson(data.toUtf8());
+    auto jsonObj = jsonDoc.object();
+
+    if (jsonObj.contains("type")) {
+        QJsonArray typesArray = jsonObj["type"].toArray();
+        for (const auto& type : typesArray) {
+            mSupportedAuthTypes.append(type.toString());
+        }
+    }
+
+#if defined(DEBUG_GMCP_AUTHENTICATION)
+    qDebug() << "Supported auth types:" << mSupportedAuthTypes;
+#endif
+}
+
+void GMCPAuthenticator::sendCredentials()
+{
+    auto character = mpHost->getLogin();
+    auto password = mpHost->getPass();
+    QJsonObject credentials;
+    if (!character.isEmpty() && !password.isEmpty()) {
+        credentials["account"] = character;
+        credentials["password"] = password;
+    }
+    QJsonDocument doc(credentials);
+    QString gmcpMessage = doc.toJson(QJsonDocument::Compact);
+
+    std::string output;
+    output += TN_IAC;
+    output += TN_SB;
+    output += OPT_GMCP;
+    output += "Char.Login.Credentials ";
+    output += mpHost->mTelnet.encodeAndCookBytes(gmcpMessage.toStdString());
+    output += TN_IAC;
+    output += TN_SE;
+
+    // Send credentials to server
+    mpHost->mTelnet.socketOutRaw(output);
+#if defined(DEBUG_GMCP_AUTHENTICATION)
+    qDebug() << "Sent GMCP credentials";
+#endif
+}
+
+
+void GMCPAuthenticator::handleAuthResult(const QString& data)
+{
+    auto doc = QJsonDocument::fromJson(data.toUtf8());
+    auto obj = doc.object();
+
+    bool success = obj[qsl("success")].toBool();
```

Comment:
Also some game drivers can parse JSON for `true` or `false`, they may not be able to put those values into an array or mapping. Currently, when `"true"` or `"false"` is passed as the value for `success` it is resolving as `false`.  Suggesting a change to resolve this:
```
    auto result = obj[qsl("success")];
    bool success = (result.isBool() && result.toBool()) || (result.isString() && result.toString() == "true");
```

Reply:
Yes, thank you.